two_factor/views/utils.py

```diff
@@ -194,3 +197,57 @@ def simple_decorator(View):
         View.dispatch = method_decorator(function_decorator)(View.dispatch)
         return View
     return simple_decorator
+
+
+def get_remember_device_cookie(user_pk, password_hash, otp_device_id):
+    """
+    Compile a signed cookie from user_pk, password_hash and otp_device_id,
+    but only return the hashed and signatures and omit the data.
+
+    The cookie is composed of 3 parts:
+    1. A weakly hashed value of otp_device_id.
+    2. A timestamp of signing.
+    3. A hmac signature that signs user_pk, password_hash, otp_device_id and the timestamp.
```

Comment:
I'm not keen on using the password hash as this data will be left in the user's browser. Is there a particular reason it's being used?

Reply:
I would prefer a solution without using the password hash as well. I included it to archive that all remembered devices become invalid once a the user changed his password. I think this is a good safety measure, as users (hopefully) reset their passwords after a breach but probably don't care about remembered devices.

I had a bad feeling as well storing the password hash in the user's browser. This is why I am only storing the signature and omit the data. A couple of lines below the data is cleaned from the cookie

Reply:
Everything but this question is fixed. 😄

Reply:
The signature still depends on the password hash, so there's always a chance that there could be some information leakage.

How about generating a random number that is stored in the cache for each user? It could then be deleted/regenerated when the user changes their password. You'd have to check this via the `pre_save` signal, something like:

```
@receiver(pre_save, sender=get_user_model())                                                                                                                                                                                                   
def user_updated_password(sender, **kwargs):                                                                                                                                                                                                   
    user = kwargs.get('instance', None)                                                                                                                                                                                                        
    new_pass = user.password                                                                                                                                                                                                                   
    try:                                                                                                                                                                                                                                       
        old_pass = get_user_model().objects.get(pk=user.pk).password                                                                                                                                                                           
    except get_user_model().DoesNotExist:                                                                                                                                                                                                      
        old_pass = None                                                                                                                                                                                                                        
    if old_pass != new_pass:                                                                                                                                                                                                                   
        # invalidate cache    
```

Reply:
I agree. Better be save than sorry.

What I like about this solution is far, is that the persistent store does not need to be changed. I guess for a caching solution we would have to alter the model of some sort?

Another approach would be to use only the salt of the encrypted password as part of encrypted data. What do you think?

Reply:
I just modified the code accordingly. Looking forward to your opinion.

Reply:
I just stumbled upon the PasswordResetTokenGenerator which also uses the password hash to create as hash sum and send this value to the user.

https://github.com/django/django/blob/da4923ea87124102aae4455e947ce24599c0365b/django/contrib/auth/tokens.py#L91

So I guess I theory it should be save to use hash the password hash. Just to be save we could implement it a bit more similar to the PasswordResetTokenGenerator. What do you think?

Reply:
With regards to using the (plaintext!) salt rather than the password hash: that's a huge no-no.

I still don't feel super comfortable with using the password hash either, but `TimestampSigner` is using all the same things that `PasswordResetTokenGenerator` does so :woman_shrugging:

So: revert your last two commits and we'll carry on from there :smile_cat:

Reply:
I went back to the old technique but removed the dependancy to `TimestampSigner`. That class was not a good fit as we do not want to store the whole signed value in a cookie. Instead I used PasswordResetTokenGenerator as a basis.

In the next version Django will move from sha1 to sha256. To increase security and prevent legacy problems I already rely on the new algorithm.

Feel free to take it from here... :-D